cc_corpus/download.py

```diff
@@ -3,94 +3,121 @@
 
 """Network-related functionality."""
 
+import boto3
 import logging
 import time
+from typing import Any, Union
 
 import requests
 from requests_toolbelt.multipart import decoder
 
 
 WARC_BASE_URI = 'https://ds5q9oxwqwsfj.cloudfront.net/'
+BUCKET_NAME = 'commoncrawl'
 
 
 class DownloadError(Exception):
     """Exception raised on if a download fails."""
     pass
 
 
-def download_ranges(url: str,
+def download_ranges(url_or_key: str,
                     offsets_and_lengths: list[tuple[int, int]],
                     retry_left: int,
                     delay_period: int = 1,
-                    ) -> list[bytes]:
+                    s3_download: bool = False,
+                    session: Any = None) -> Union[bytes, list[bytes]]:
     """
     Downloads a list of ranges from a URL.
 
-    :param url: the URL the byte ranges are downloaded from.
+    :param url_or_key: the URL or S3 object key of the source.
     :param offsets_and_lengths: the byte ranges to download, represented as
                                 offset-lengths pairs.
     :param retry_left: the number of retries left.
     :param delay_period: the base time unit for delay in seconds.
+    :param s3_download: whether to download via S3 or http.
+    :param session: S3 session to use. If None and we download via S3, a new
+                    session will be created.
     """
     logging.debug(
-        f'Downloading {len(offsets_and_lengths)=} ranges from {url}.')
+        f'Downloading {len(offsets_and_lengths)=} ranges from {url_or_key}.')
```

Comment:
Didn't you say that you cannot download a list of ranges from s3? So either the function should raise `ValueError` `if s3_download and len(offsets_and_lengths) > 1` or it should call an s3-specific downloader function in a loop with a single range at a time.

Reply:
Done.